File: src/main/java/com/kaankaplan/booke/business/abstracts/ReaderService.java
package com.kaankaplan.booke.business.abstracts;

import com.kaankaplan.booke.core.util.results.DataResult;
import com.kaankaplan.booke.core.util.results.Result;
import com.kaankaplan.booke.dto.BookStatusDto;
import com.kaankaplan.booke.modals.Reader;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface ReaderService {

    DataResult<Reader> saveReader(Reader reader);

    DataResult<Reader> getReaderById(String userId);
    DataResult<Reader> getReaderByEmail(String email);
    DataResult<List<Reader>> getReadersFollows(String userId);
    DataResult<List<Reader>> getReadersFollowers(String userId);

    Result follow(String userId, String wantToFollowUserId);

    Result unfollow(String userId, String wantToUnfollowUserId);

    Result addPostToReader(String userId, String postId);

    DataResult<BookStatusDto> getBookCurrentStatusForReader(String readerId, String bookId);

    Result addBookIntoCurrentlyReadings(String readerId, String bookId);

    Result addBookIntoWantToReads(String readerId, String bookId);

    Result addBookIntoReads(String readerId, String bookId);

    DataResult<List<Reader>> searchReaderAutocomplete(String fullName);

    Result updateProfileImage(String userId, MultipartFile multipartFile);

    Result startReadingChallenge(String userId, int target);

}

File: src/main/booke-mobile-app/services/ReaderService.js
import Interceptor from "../utils/interceptors/Interceptor";

export default class ReaderService {
    apiUrl = "reader/";
    axiosInstance;

    constructor() {
        this.axiosInstance = new Interceptor().getInstance();
    }

    getReaderByEmail(email) {
        return this.axiosInstance.get(this.apiUrl + "getReaderByEmail/" + email);
    }
    
    getReaderById(readerId) {
        return this.axiosInstance.get(this.apiUrl + "getReaderById/" + readerId);
    }

    getReaderFollows(readerId) {
        return this.axiosInstance.get(this.apiUrl + "getReaderFollows/" + readerId);
    }

    getReaderFollowers(readerId) {
        return this.axiosInstance.get(this.apiUrl + "getReaderFollowers/" + readerId);
    }

    getBookStatusForUser(userId, bookId) {
        return this.axiosInstance.get(this.apiUrl + "getBookStatus/" + userId + "/" + bookId);
    }

    addBookIntoWantToReads(userId, bookId) {
        return this.axiosInstance.post(this.apiUrl + "addWantToRead/" + userId + "/" + bookId, {});
    }

    addBookIntoCurrentlyReadings(userId, bookId) {
        return this.axiosInstance.post(this.apiUrl + "addCurrentlyBook/" + userId + "/" + bookId, {});
    }

    addBookIntoReads(userId, bookId) {
        return this.axiosInstance.post(this.apiUrl + "addRead/" + userId + "/" + bookId, {});
    }

    startReadingChallenge(userId, target) {
        return this.axiosInstance.post(this.apiUrl + "startChallenge/" + userId + "/" + target, {})
    }

    searchReader(name) {
        return this.axiosInstance.get(this.apiUrl + "search/" + name);
    }

    follow(userId, wantToFollowUserId) {
        return this.axiosInstance.post(this.apiUrl + "follow/" + userId + "/" + wantToFollowUserId, {});
    }

    unfollow(userId, wantToUnfollowUserId) {
        return this.axiosInstance.post(this.apiUrl + "unfollow/" + userId + "/" + wantToUnfollowUserId, {});
    }

    updateProfileImage(userId, formData) {
        return this.axiosInstance.post(this.apiUrl + "updateProfileImage/" + userId, formData, {
            headers: {
                "Content-Type": "multipart/form-data",
            }
        })
    }
}